### PR TITLE
[node] Add v22 typings

### PR DESCRIPTION
**Pull request by Semigradsky**
It is a small step to v22 typings.

- Run scripts/new-version 22
- Replace `v20.#.#` to `v22.x` in links

----------------------------------------------------

https://github.com/nodejs/node/releases/tag/v22.0.0

- esm: drop support for import assertions
- stream: bump default highWaterMark
- fs: expose glob and globSync
- crypto: move createCipher and createDecipher to eol
- fs: runtime deprecate fs.Stats constructor

### CODE REVIEW

**Comment by Semigradsky** [comment]
There are errors for TS@5.6 like
```
expected type to be:
  Generator<unknown, any, unknown>
got:
  Generator<unknown, any, any>
```

**Comment by jakebailey** [comment]
You'll need to make changes like #69632 for this to work out.

**Comment by jakebailey** [comment]
Is someone from the Node side going to review it?

**Comment by Renegade334** [comment]
Leading from an issue raised in #definitely-typed today, it emerged that although the `suite` definitions (v20/v22) were added to test.d.ts under #69508, they were never actually exported from the module, and there were no DT tests to pick this up. This seems like an appropriate PR to include these.

Drafted fc95df9fae4bafa44ac888e589868f7481c8d66d – feel free to cherry-pick or modify at leisure. I haven't run it through CI.

File: types/node/test/fs.ts [hunk 1]
```diff
@@ -896,3 +897,27 @@ const anyStatFs: fs.StatsFs | fs.BigIntStatsFs = fs.statfsSync(".", { bigint: Ma
...
+    glob("**/*.js", { withFileTypes: true }, (err, matches) => {
+        matches; // $ExpectType Dirent[]
+    });
+
+    for (const entry of globSync('**/*.js')) {
+        entry; // $ExpectType string
+    }
+    for (const entry of globSync('**/*.js', { withFileTypes: true })) {
+        entry; // $ExpectType Dirent
+    }
```
Comment (Renegade334) [inline types/node/test/fs.ts:922]:
```suggestion
    for (const entry of globSync('**/*.js')) {
        entry; // $ExpectType string
    }
    for (const entry of globSync('**/*.js', { cwd: '/' })) {
        entry; // $ExpectType string
    }
    for (const entry of globSync('**/*.js', { withFileTypes: true })) {
        entry; // $ExpectType Dirent
    }
    for (const entry of globSync('**/*.js', { withFileTypes: Math.random() > 0.5 })) {
        entry; // $ExpectType Dirent | string
    }
```

As above.

**Review by Renegade334 — CHANGES_REQUESTED** [review]
Just a few suggestions related to `glob`.

**Review by Renegade334 — APPROVED** [review]

File: types/node/fs/promises.d.ts [hunk 2]
```diff
@@ -1239,6 +1242,21 @@ declare module "fs/promises" {
...
+    /**
+     * Retrieves the files matching the specified pattern.
+     */
+    function glob<Options extends GlobOptions>(
+        pattern: string | string[],
+        opt: Options,
+    ): Options extends GlobOptionsWithFileTypes
+        ? AsyncIterableIterator<Dirent>
+        : Options extends GlobOptionsWithoutFileTypes
+            ?  AsyncIterableIterator<string>
```
Comment (Renegade334) [inline types/node/fs/promises.d.ts:1258]:
```suggestion
            ? AsyncIterableIterator<string>
```
😉

File: types/node/fs/promises.d.ts [hunk 3]
```diff
@@ -1239,6 +1242,21 @@ declare module "fs/promises" {
...
+            : AsyncIterableIterator<Dirent | string>;
```
Comment (Renegade334) [inline types/node/fs/promises.d.ts:1259]: Semantically this should be `AsyncIterableIterator<Dirent> | AsyncIterableIterator<string>`, even though in practice it's impossible to tell the difference without consuming the iterator.

**Comment by Semigradsky** [comment]
@Renegade334 great thanks for comments! 👍

**Comment by jakebailey** [comment]
I know this PR deletes Node v16, but I believe that due to some iterator changes in 5.6, it's likely the changes we had to make for 5.6 will need to be revised in the Node types before 5.6 is totally released; would you be open to leaving v16 behind for now, just to make sure that we can keep it patched, at least until 5.6 is fully released?

**Comment by Semigradsky** [comment]
Ok I will return v16, no probs 😸

File: types/node/package.json [hunk 4]
```diff
@@ -1,15 +1,15 @@
 {
     "private": true,
     "name": "@types/node",
-    "version": "20.14.9999",
+    "version": "22.0.9999",
```
Comment (jakebailey) [inline types/node/package.json:4]: Just checking, this says `22.0`; should it be a patch like `22.5`? Or does this really only contain things added in 22.0?

File: types/node/module.d.ts [hunk 5]
```diff
@@ -156,10 +154,6 @@ declare module "module" {
              * Export conditions of the relevant `package.json`
              */
             conditions: string[];
-            /**
-             * @deprecated Use `importAttributes` instead
-             */
-            importAssertions: ImportAttributes;
```
Comment (jakebailey) [inline types/node/module.d.ts:162]: The docs don't say this has been removed, and I can see the runtime code still using this; is this removal intentional?

Comment (jakebailey) [reply types/node/module.d.ts:162]: https://github.com/nodejs/node/pull/52104 dropped support for the syntax and dynamic import's second parameter, but apparently did not remove it fully (e.g. https://github.com/nodejs/node/blob/2d1b4a8cf7ce875f0c458941bb40ba6fbdccd42f/lib/internal/modules/esm/load.js#L112); not sure where to draw the line here but I guess it's probably okay to remove it and see what happens?

**Comment by jakebailey** [comment]
Seems alright to me (I gave it a proper scan), just a couple of questions about the bits that were removed.

File: types/node/package.json (hunk 4 above)
Comment (jakebailey) [reply types/node/package.json:4]: I checked the PR and I think it does all check out for 22.0.

**Review by jakebailey — APPROVED** [review]

File: types/node/module.d.ts (hunk 5 above)
Comment (Semigradsky) [reply types/node/module.d.ts:162]: New major version looks like a good time to remove deprecated  things 😄

**Comment by Semigradsky** [comment]
Ready to merge

**Comment by egoroof** [comment]
sorry, but where is `WebSocket` types? https://nodejs.org/docs/latest/api/globals.html#websocket

**Comment by Renegade334** [comment]
> where is `WebSocket` types? 

The global `WebSocket` object is a web API defined in TypeScript's DOM library.

Websockets are also exposed via the `node:http` module as of 22.5, but the DefinitelyTyped definitions need to be updated in sequence for each version, so this will take some time to land.

**Comment by egoroof** [comment]
> The global WebSocket object is a web API defined in TypeScript's DOM library.

I meant this. But even with
```
"compilerOptions": {
    "lib": ["ESNext"]
}
```
Typescript 5.5.4 doesn't know about `WebSocket`

**Comment by egoroof** [comment]
> The global WebSocket object is a web API defined in TypeScript's DOM library.

I meant this. But with
```
"compilerOptions": {
  "lib": ["ESNext"],
}
```
Typescript 5.5.4 doesn't know about `WebSocket`

**Comment by egoroof** [comment]
Ahh I found https://github.com/microsoft/TypeScript/issues/53440 and https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924

Also I found that for example @types/node `fetch` type doesn't know about `cache` init param.. Now I am confused

**Comment by egoroof** [comment]
undici exposes `WebSocket` types, they should be included like https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66824/ ?

**Comment by Semigradsky** [comment]
PR for WebSocket: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70166

**Pull request merged**